lib/setup.c

```diff
@@ -921,6 +926,40 @@ ___SIZE_TS requested_words_still;)
 }
 
 
+___EXP_FUNC(___BOOL, ___fdset_resize)
+   ___P((int fd1,
+         int fd2),
+        (fd1,
+         fd2)
+int fd1;
+int fd2;)
+{
+#ifdef USE_select_or_poll
+
+  ___processor_state ___ps = ___PSTATE;
+  int fd = (fd2 > fd1) ? fd2 : fd1;
+
+  if (fd < ___ps->os.fdset.size)
+    return 0;
+
+  ___fdset_resize_heap_overflow_clear ();
+
+  ___sync_op_struct sop;
+
+  sop.op = OP_FDSET_RESIZE;
+  sop.arg[0] = fd;
+
+  on_all_processors (___PSP &sop);
+
+  return ___fdset_resize_heap_overflow ();
```

Comment:
This kind of situation can be solved with a barrier synchronisation, and that is how the garbage collector deals with different "phases" of the parallel GC that need to not overlap between processors.  The idea is to

```
#ifdef ___SINGLE_THREADED_VMS
#define BARRIER()
#else
#define BARRIER() barrier_sync_noop (___PSPNC)
#endif
```

and then in the code for the resize, code like this:

```
if (___PROCESSOR_ID(___ps,___VMSTATE_FROM_PSTATE(___ps)) == 0)
  fdset_heap_overflow = 0;

BARRIER();

... do the resize ...

BARRIER();

if (fdset_heap_overflow)
  ...;
```